internal/events/persist_batch.go

```diff
@@ -31,63 +31,94 @@ func (em *eventManager) persistBatchFromBroadcast(ctx context.Context /* db TX c
-		return false, nil // This is not retryable. skip this batch
+		return nil, false, nil // This is not retryable. skip this batch
 	}
 
 	// Set confirmed on the batch (the messages should not be confirmed at this point - that's the aggregator's job)
-	batch.Confirmed = now
+	persistedBatch, _ = batch.Confirmed()
+	manifestHash := fftypes.HashString(persistedBatch.Manifest.String())
+
+	// Verify the hash calculation.
+	if !manifestHash.Equals(batch.Hash) {
+		// To cope with existing batches written by v0.13 and older environments, we have to do a more expensive
+		// hashing of the whole payload before we reject.
+		if batch.Payload.Hash().Equals(batch.Hash) {
+			l.Infof("Persisting migrated batch '%s'. Hash is a payload hash: %s", batch.ID, batch.Hash)
+		} else {
+			l.Errorf("Invalid batch '%s'. Hash does not match payload. Found=%s Expected=%s", batch.ID, manifestHash, batch.Hash)
+			return nil, false, nil // This is not retryable. skip this batch
+		}
+	}
 
-	// Upsert the batch itself, ensuring the hash does not change
-	err = em.database.UpsertBatch(ctx, batch)
+	// Upsert the batch
+	err = em.database.UpsertBatch(ctx, persistedBatch)
 	if err != nil {
 		if err == database.HashMismatch {
 			l.Errorf("Invalid batch '%s'. Batch hash mismatch with existing record", batch.ID)
-			return false, nil // This is not retryable. skip this batch
+			return nil, false, nil // This is not retryable. skip this batch
 		}
 		l.Errorf("Failed to insert batch '%s': %s", batch.ID, err)
-		return false, err // a persistence failure here is considered retryable (so returned)
+		return nil, false, err // a persistence failure here is considered retryable (so returned)
 	}
 
+	valid, err = em.persistBatchContent(ctx, batch)
+	if err != nil || !valid {
+		return nil, valid, err
+	}
+	return persistedBatch, valid, err
+}
+
+func (em *eventManager) persistBatchContent(ctx context.Context, batch *fftypes.Batch) (valid bool, err error) {
+
 	optimization := em.getOptimization(ctx, batch)
 
 	// Insert the data entries
+	dataByID := make(map[fftypes.UUID]*fftypes.Data)
 	for i, data := range batch.Payload.Data {
-		if err = em.persistBatchData(ctx, batch, i, data, optimization); err != nil {
-			return false, err
+		if valid, err = em.persistBatchData(ctx, batch, i, data, optimization); !valid || err != nil {
+			return valid, err
 		}
+		dataByID[*data.ID] = data
 	}
 
 	// Insert the message entries
 	for i, msg := range batch.Payload.Messages {
 		if valid, err = em.persistBatchMessage(ctx, batch, i, msg, optimization); !valid || err != nil {
 			return valid, err
 		}
+		dataInBatch := true
+		msgData := make(fftypes.DataArray, len(msg.Data))
+		for di, dataRef := range msg.Data {
+			msgData[di] = dataByID[*dataRef.ID]
+			if msgData[di] == nil || !msgData[di].Hash.Equals(dataRef.Hash) {
+				log.L(ctx).Debugf("Message '%s' in batch '%s' - data not in-line in batch id='%s' hash='%s'", msg.Header.ID, batch.ID, dataRef.ID, dataRef.Hash)
```

Comment:
Should this be higher than debug? Is this an expected situation?

Reply:
The architecture prior to this code change allows it.
e.g. you could send some data to a party, then send a message referring to that data, without sending that data again.